---
title: "ADA Final Project"
author: "Grace Musser and Erin Keenan-Early"
date: "4/30/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Preliminary Analyses

##Phylogenetic data matrix creation
Here, we used a combination of 13 nuclear and mitochondrial genes for avian taxa.  

##Bayesian analysis

###Bayesian analysis overview

###Models of evolution
We ran the phylogenetic analyses using Bayesian methods using MrBayes on TACC. We ran the analysis using the GTR (Generalised time reversible) model with gamma-distributed rate variation across sites. If we did not include the gamma model, it would be assumed that all sites are evolving at equal rates. The GTR model is most often used in current phylogenetic studies as it is the most complex model currently available for molecular evolution. It comprises 64 nested models with various combinations of parameters for DNA site substitution and has the highest degrees of freedom out of all previously published models of molecular evolution.

#Downstream analyses in R

##Tree extraction 

First, load the required libraries:
```{r}
library(ape)
library(phytools)
library(BAMMtools)
library(readxl)
library(curl)
```

And the color palette:
```{r}
library(gameofthrones)
```


You can grab the data here:

Body mass data: (may need to be downloaded directly as excel file)
```{r}
f<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/ONLYGruiBodyMassData.csv")
GRUIMassData<- read.csv(f, header = TRUE, sep = ",", stringsAsFactors = TRUE)
head(GRUIMassData)
```


Posterior tree set:
```{r}
posteriors <- curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/posterior_tree_set.nex")
```

BAMM mcmcout file:
```{r}
f <- curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/mcmc_out.txt")
mcmcout <- read.csv(f, header = TRUE, sep = ",", stringsAsFactors = TRUE)
head(mcmcout)
```

Newick tree:
```{r}
my_Newick_tree<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/MyNewickTreefile.tre")
```


BAMM event data (may need to be downloaded and pulled from working directory)
```{r}
sevent<-curl("https://raw.githubusercontent.com/gmusser553/ADA_Final_Project/master/sevent_data.txt")
```


Extract a tree from a posterior tree set (in this case it is in nexus format). Then assign a tree from this posterior tree set to a variable. In this case we picked tree 42:
```{r}
#grui.tree<-read.nexus(file="posterior_tree_set.nex") 

grui.tree<-read.nexus(file=posteriors) 

subsettree<-grui.tree[[42]]

```

Plot the exracted tree in a fan style to check that this worked:
```{r}
plotTree(subsettree,type="fan",ftype="i", color=got(150, option = "Targaryen2"), fsize=0.3)
```

##Ancestral State Reconstruction

Here, we read in body mass data and drop names that do not match. We also take the log of the mean body mass data so that it will be easier to visalize after performing ancestral state reconstruction and plotting it on the tree.

```{r}
GRUIMassData <- read_excel("ONLYGruiBodyMassData.xlsx") 



meanmass<-GRUIMassData$LogMean

names(meanmass)<-GRUIMassData$`Species Name`
xxx <- na.omit(meanmass)

v <- subsettree$tip.label %in% names(xxx)
drop <- subsettree$tip.label[v==FALSE]
d2<-drop.tip(subsettree, drop)
```


Perform ancestral state reconstruction using fastAnc:

```{r}
fit<-fastAnc(tree=d2, xxx, vars=TRUE, CI=TRUE)
fit

```

Use contMap to create a tree with the ancestral state reconstructions plotted onto it. We assign it to a variable here, but you can also use it alone to create a tree image. 

contMap takes the tree as argument x, the continuous traits to be input and from which ancestral state reconstruction can be estimated, and has several additional options that can be employed. 

Here, we decrease line width for branches since we have large taxon sampling (lwd) and increase resolution of the ancestral state reconstruction (res). fsize can be used in plot or contMap and determines font size.

```{r}
obj<-contMap(d2,xxx,plot=FALSE, res=1000, lwd=.5)

```

The setMap function allows us to change the color palette of the tree.

```{r}
obj<-setMap(obj, colors=got(128, option = "Daenerys"))
```

```{r}
plot(obj, fsize=.2)
```


We can also plot it in a fan style. Fan styles are typically used for datasets with large taxon sampling.

```{r}
obj<-setMap(obj, colors=got(256, option = "Targaryen2"))

plot(obj,type="fan",legend=0.7*max(nodeHeights(d2)),
     fsize=.5)
```

We can also plot a phenogram. A phenogram depicts taxonomic relationships based on overall similarity of character states without takiing evolutionary history or weighting/ordering of characetrs. 

```{r}
phenogram(d2,xxx,fsize=0.2,spread.costs=c(1,0))
```



##BAMM analyses and plotting in R:


```{r}
#first use BAMM in terminal-must be Newick notation
#check that tree is ultrametric and all tree lengths are greater than 0:
is.ultrametric(d2) #check if ultrametric
min(d2$edge.length) #check minimum branch length

#After running BAMM, load mcmc_out.txt or equivalent file into R to test convergence using 'coda' package:
#mcmcout <- read.csv("mcmc_out.txt", header=T)
plot(mcmcout$logLik ~ mcmcout$generation)

#next: discard burnin
burnstart <- floor(0.1 * nrow(mcmcout))
postburn <- mcmcout[burnstart:nrow(mcmcout), ]

#check the effective sample sizes of the log-likelihood and the number of shift events present in each 
#you want this to be at least 200
#effectiveSize(postburn$N_shifts)
#effectiveSize(postburn$logLik)

#tree <- read.tree("MyNewickTreefile.tre")
tree<-read.tree(my_Newick_tree)
edata <- getEventData(tree, eventdata = "sevent_data.txt", burnin=0.1)

shift_probs <- summary(edata) #create dataframe of posterior probabilities
#visualizing mean, model-averaged diversification rates at any point along every branch of a phylogenetic tree:
plot.bammdata(edata, lwd=2, legend=T)

index <- 25
e2 <- subsetEventData(edata, index = index)
plot.bammdata(e2, lwd=2, legend=TRUE)
addBAMMshifts(e2, cex=2)

```




#Phylogenetically informed correlation of BAMM rates and body mass
```{r}
D.scores <- GRUIMassData$LogMean
names(D.scores) <- GRUIMassData$`Species Name`

is.D.scores.in.edata <- names(D.scores) %in% edata$tip.label
  
D.scores.trim <- D.scores[is.D.scores.in.edata==TRUE]

  is.edata.in.D.scores <- edata$tip.label %in% names(D.scores.trim)

  keep <- edata$tip.label[is.edata.in.D.scores == TRUE]
  
  edata.subset <- subtreeBAMM(edata, tips=keep)
  
 cor.result <- traitDependentBAMM(edata.subset, traits=D.scores.trim,
                                     reps=1000, return.full = TRUE,
                                    method="s",logrates = TRUE,
                                    two.tailed = TRUE)
 
 #correlation
hist(cor.result$estimate, xlim=c(-1,1))
hist(cor.result$obs.corr, xlim=c(-1,1))
cor.result$estimate
cor.result$p.value
```

```{r}

```

